webextension/background.js

```diff
@@ -333,30 +347,62 @@ const themeManager = {
 };
 
 const tabPageCounter = {
-  counter: {},
-
-  initTabCounter(tab) {
-    if (tab.id in this.counter) {
+  counters: {},
+
+  initTabCounter(tab, why = "user-activated-tab") {
+    // When the user becomes active,
+    // we ONLY need to initialize the activity counter
+    if (tab.id in this.counters && why === "user-became-active") {
+      this.counters[tab.id].activity = {
+        "cookieStoreId": tab.cookieStoreId,
+        "pageRequests": 0
+      };
+      return;
+    }
+    if (tab.id in this.counters) {
       return;
     }
-    this.counter[tab.id] = {
+    this.counters[tab.id] = {};
+    this.counters[tab.id].tab = {
+      "cookieStoreId": tab.cookieStoreId,
+      "pageRequests": 0
+    };
+    this.counters[tab.id].activity = {
       "cookieStoreId": tab.cookieStoreId,
       "pageRequests": 0
     };
   },
 
-  sendTabCountAndDelete(tabId) {
-    browser.runtime.sendMessage({
-      method: "sendTelemetryPayload",
-      event: "page-requests-completed-per-tab",
-      userContextId: this.counter[tabId].cookieStoreId,
-      pageRequestCount: this.counter[tabId].pageRequests
-    });
-    delete this.counter[tabId];
+  sendTabCountAndDelete(tabId, why = "user-closed-tab") {
+    if (why === "user-closed-tab") {
+      browser.runtime.sendMessage({
+        method: "sendTelemetryPayload",
+        event: "page-requests-completed-per-tab",
+        userContextId: this.counters[tabId].tab.cookieStoreId,
+        pageRequestCount: this.counters[tabId].tab.pageRequests
+      });
+      // When we send the ping because the user closed the tab,
+      // delete both the 'tab' and 'activity' counters
+      delete this.counters[tabId];
+    } else if (why === "user-went-idle") {
+      browser.runtime.sendMessage({
+        method: "sendTelemetryPayload",
+        event: "page-requests-completed-per-activity",
+        userContextId: this.counters[tabId].activity.cookieStoreId,
+        pageRequestCount: this.counters[tabId].activity.pageRequests
+      });
+      // When we send the ping because the user went idle,
+      // only reset the 'activity' counter
+      this.counters[tabId].activity = {
```

Comment:
Well-spotted.

You're right there shouldn't be any counts in between. But, I'll call it defensive coding for a future when someone might call `initTabCounter(tab, "user-became-active")` without a previous call to `sendTabCountAndDelete(tab, "user-went-idle").

I just caught a bug in our Telemetry data because [we call `showTabs` from within `openTab`](https://github.com/mozilla/testpilot-containers/blob/3e657a2e8d952da30c963ae929bc8a774db9a425/index.js#L861), resulting in duplicate `show-tabs` event pings that I had to filter out in the SQL query. :cry: